src/proof/lrat/lrat_proof.cpp

```diff
@@ -0,0 +1,199 @@
+/*********************                                                        */
+/*! \file lrat_proof.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief DRAT Proof Format
+ **
+ ** Defines deserialization for DRAT proofs.
+ **/
+
+#include "proof/lrat/lrat_proof.h"
+
+#include "base/output.h"
+#include "base/cvc4_assert.h"
+
+#include <algorithm>
+#include <cstdlib>
+#include <iostream>
+#include <fstream>
+#include <unordered_map>
+
+namespace CVC4 {
+namespace proof {
+namespace lrat {
+
+using prop::SatClause;
+using prop::SatLiteral;
+using prop::SatVariable;
+
+LRATProof::LRATProof(
+    const std::unordered_map<ClauseId, prop::SatClause*>& usedClauses,
+    const std::vector<ClauseId>& clauseOrder,
+    const std::string& dratBinary)
+{
+  std::ostringstream cmd;
+  char formulaFilename[] = "/tmp/cvc4-dimacs-XXXXXX";
+  char dratFilename[] = "/tmp/cvc4-drat-XXXXXX";
+  char lratFilename[] = "/tmp/cvc4-lrat-XXXXXX";
+  int r;
+  r = mkstemp(formulaFilename);
+  Assert(r > 0);
+  r = mkstemp(dratFilename);
+  Assert(r > 0);
+  r = mkstemp(lratFilename);
+  Assert(r > 0);
+  std::ofstream formStream(formulaFilename);
+  size_t maxVar = 0;
+  for (auto& c : usedClauses)
+  {
+    for (auto l : *(c.second))
+    {
+      if (l.getSatVariable() + 1 > maxVar)
+      {
+        maxVar = l.getSatVariable() + 1;
+      }
+    }
+  }
+  formStream << "p cnf " << maxVar << " " << usedClauses.size() << '\n';
+  for (auto ci : clauseOrder)
+  {
+    auto iterator = usedClauses.find(ci);
+    Assert(iterator != usedClauses.end());
+    for (auto l : *(iterator->second))
+    {
+      if (l.isNegated())
+      {
+        formStream << '-';
+      }
+      formStream << l.getSatVariable() + 1 << " ";
+    }
+    formStream << "0\n";
+  }
+  formStream.close();
+
+  std::ofstream dratStream(dratFilename);
+  dratStream << dratBinary;
+  dratStream.close();
+
+  // TODO(aozdemir) Add invocation of DRAT trim, once I get CMake to bundle it
+  // into CVC4 correctly.
+  Unimplemented();
+
+  std::ifstream lratStream(lratFilename);
+  LRATProof lratProof(lratStream);
+  remove(formulaFilename);
+  remove(dratFilename);
+  remove(lratFilename);
+  *this = std::move(lratProof);
```

Comment:
Instead of doing this (I am actually not quite sure whether the C++ standard allows that or not), I'd suggest using a static class method `static std::unique_ptr<LratProof> LratProof::fromDratProof(...);`.

Reply:
I like the idea of a static constructor-like function (incidentally, that's how **all** Rust constructors work). How about just returning an `LratProof`, and letting return-value-optimization do its work?

Reply:
Yes, that's actually even better.